File: backend/src/schemas/transaction.py
# backend/src/schemas/transaction.py

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum


class TransactionType(str, Enum):
    PURCHASE = "PURCHASE"
    CONSUMPTION = "CONSUMPTION"
    ADJUSTMENT = "ADJUSTMENT"
    EXPIRED = "EXPIRED"


class TransactionBase(BaseModel):
    lebensmittel_id: int
    transaction_type: TransactionType
    quantity_change: int = Field(..., description="Positive for additions, negative for reductions")
    reason: Optional[str] = None


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(BaseModel):
    reason: Optional[str] = None

    model_config = ConfigDict(
        populate_by_name=True
    )


class TransactionRead(TransactionBase):
    id: int
    quantity_before: Optional[int] = None
    quantity_after: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True
    )


class TransactionWithLebensmittel(TransactionRead):
    lebensmittel_name: str
    lebensmittel_kategorie: Optional[str] = None


# Statistics schemas
class CategoryStatistics(BaseModel):
    kategorie: str
    total_purchases: int
    total_consumption: int
    net_change: int
    transaction_count: int


class MonthlyStatistics(BaseModel):
    year: int
    month: int
    total_purchases: int
    total_consumption: int
    net_change: int
    transaction_count: int


class LebensmittelStatistics(BaseModel):
    lebensmittel_id: int
    lebensmittel_name: str
    kategorie: Optional[str] = None
    total_purchases: int
    total_consumption: int
    net_change: int
    transaction_count: int
    last_transaction: Optional[datetime] = None


class StatisticsOverview(BaseModel):
    total_transactions: int
    total_purchases: int
    total_consumption: int
    most_purchased_category: Optional[str] = None
    most_consumed_category: Optional[str] = None
    categories: list[CategoryStatistics]
    monthly: list[MonthlyStatistics]
    top_items: list[LebensmittelStatistics]

File: android-client/app/src/main/java/com/foodtrack/app/data/model/Transaction.kt
package com.foodtrack.app.data.model

import com.google.gson.annotations.SerializedName

enum class TransactionType {
    @SerializedName("PURCHASE")
    PURCHASE,
    
    @SerializedName("CONSUMPTION")
    CONSUMPTION,
    
    @SerializedName("ADJUSTMENT")
    ADJUSTMENT,
    
    @SerializedName("EXPIRED")
    EXPIRED
}

data class Transaction(
    val id: Int,
    @SerializedName("lebensmittel_id") val lebensmittelId: Int,
    @SerializedName("transaction_type") val transactionType: TransactionType,
    @SerializedName("quantity_change") val quantityChange: Int,
    @SerializedName("quantity_before") val quantityBefore: Int?,
    @SerializedName("quantity_after") val quantityAfter: Int?,
    val reason: String?,
    @SerializedName("created_at") val createdAt: String
)

data class TransactionCreate(
    @SerializedName("lebensmittel_id") val lebensmittelId: Int,
    @SerializedName("transaction_type") val transactionType: TransactionType,
    @SerializedName("quantity_change") val quantityChange: Int,
    val reason: String?
)

// Statistics models
data class CategoryStatistics(
    val kategorie: String,
    @SerializedName("total_purchases") val totalPurchases: Int,
    @SerializedName("total_consumption") val totalConsumption: Int,
    @SerializedName("net_change") val netChange: Int,
    @SerializedName("transaction_count") val transactionCount: Int
)

data class MonthlyStatistics(
    val year: Int,
    val month: Int,
    @SerializedName("total_purchases") val totalPurchases: Int,
    @SerializedName("total_consumption") val totalConsumption: Int,
    @SerializedName("net_change") val netChange: Int,
    @SerializedName("transaction_count") val transactionCount: Int
)

data class LebensmittelStatistics(
    @SerializedName("lebensmittel_id") val lebensmittelId: Int,
    @SerializedName("lebensmittel_name") val lebensmittelName: String,
    val kategorie: String?,
    @SerializedName("total_purchases") val totalPurchases: Int,
    @SerializedName("total_consumption") val totalConsumption: Int,
    @SerializedName("net_change") val netChange: Int,
    @SerializedName("transaction_count") val transactionCount: Int,
    @SerializedName("last_transaction") val lastTransaction: String?
)

data class StatisticsOverview(
    @SerializedName("total_transactions") val totalTransactions: Int,
    @SerializedName("total_purchases") val totalPurchases: Int,
    @SerializedName("total_consumption") val totalConsumption: Int,
    @SerializedName("most_purchased_category") val mostPurchasedCategory: String?,
    @SerializedName("most_consumed_category") val mostConsumedCategory: String?,
    val categories: List<CategoryStatistics>,
    val monthly: List<MonthlyStatistics>,
    @SerializedName("top_items") val topItems: List<LebensmittelStatistics>
)
